Compilation error with GCC

Hi folks,

When fully integrating CppADCodeGen into Crocoddyl using explicit template instantiation, I noticed a compilation error with GCC compilers. I have spend a few days trying to understand the issue, but it seems something related with CppADCodeGen, and not Crocoddyl. Below, I shared the error with encounter only on GCC compiler:
```bash
In file included from /usr/share/miniconda/envs/crocoddyl/include/cppad/cg/cppadcg.hpp:122,
                 from /usr/share/miniconda/envs/crocoddyl/include/cppad/cg.hpp:18,
                 from /usr/share/miniconda/envs/crocoddyl/include/cppad/cg/support/cppadcg_eigen.hpp:18,
                 from /home/runner/work/crocoddyl/crocoddyl/include/crocoddyl/core/macros.hpp:17,
                 from /home/runner/work/crocoddyl/crocoddyl/include/crocoddyl/core/pch.hpp:30,
                 from /home/runner/work/crocoddyl/crocoddyl/build/CMakeFiles/crocoddyl.dir/cmake_pch.hxx:5,
                 from <command-line>:
/usr/share/miniconda/envs/crocoddyl/include/cppad/cg/cg.hpp:274:76: error: redefinition of 'bool __tls_guard'
  274 | thread_local std::function<std::ostream& (std::ostream&, const CG<Base>&)> CGOStreamFunc<Base>::FUNC = nullptr;
```

This error was produced by our CI in https://github.com/loco-3d/crocoddyl/actions/runs/14486448690/job/40632871703?pr=1377. Additionally, you can see that the Clang compiler works well in https://github.com/loco-3d/crocoddyl/actions/runs/14486448690/job/40632871696?pr=1377.

I would love to hear from you soon and try to fix this current issue.

Many thanks,
Carlos

I setup a minimal reproducible example:

```cpp
#include <cppad/cg.hpp>

struct A {};
struct B {};

template <typename T>
struct Wrapper {
  virtual void foo() { 
    v.FUNC;
  }
  CppAD::cg::CGOStreamFunc<T> v;
};

int main() {
  Wrapper<A> a;
  Wrapper<B> b;
  a.v.FUNC;
};
```

Build instructions (with conda package):
```bash
pixi init
pixi add cxx-compiler cppadcodegen
pixi shell
g++ test.cpp -isystem .pixi/envs/default/include -L .pixi/envs/default/lib -lcppad_lib
```

Error message:
```
In file included from .pixi/envs/default/include/cppad/cg/cppadcg.hpp:122,
                 from .pixi/envs/default/include/cppad/cg.hpp:18,
                 from test.cpp:1:
.pixi/envs/default/include/cppad/cg/cg.hpp:274:76: error: redefinition of 'bool __tls_guard'
  274 | thread_local std::function<std::ostream& (std::ostream&, const CG<Base>&)> CGOStreamFunc<Base>::FUNC = nullptr;
      |                                                                            ^~~~~~~~~~~~~~~~~~~
.pixi/envs/default/include/cppad/cg/cg.hpp:274:76: note: 'bool __tls_guard' previously declared here
.pixi/envs/default/include/cppad/cg/cg.hpp:274: confused by earlier errors, bailing out
```

--- a/include/cppad/cg/cg.hpp
+++ b/include/cppad/cg/cg.hpp
@@ -267,11 +267,14 @@ protected:
  */
 template<class Base>
 struct CGOStreamFunc {
-    static thread_local std::function<std::ostream& (std::ostream&, const CG<Base>&)> FUNC;
+    // Because of a GCC bug (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66944)
+    // FUNC_OBJ can't be a CGOStreamFunc static member.
+    // Instead, we store it as CGOStreamFunc::FUNC method static variable.
+    static std::function<std::ostream& (std::ostream&, const CG<Base>&)>& FUNC () {
+        static thread_local std::function<std::ostream& (std::ostream&, const CG<Base>&)> FUNC_OBJ = nullptr;
+        return FUNC_OBJ;
+    }
 };
-
-template<class Base>
-thread_local std::function<std::ostream& (std::ostream&, const CG<Base>&)> CGOStreamFunc<Base>::FUNC = nullptr;
 
 /**
  * Output stream operator for CG objects.
@@ -282,8 +285,8 @@ inline std::ostream& operator<<(
         std::ostream& os, //< stream to write to
         const CG<Base>& v//< vector that is output
         ) {
-    if(CGOStreamFunc<Base>::FUNC != nullptr) {
-        return CGOStreamFunc<Base>::FUNC(os, v);
+    if(CGOStreamFunc<Base>::FUNC() != nullptr) {
+        return CGOStreamFunc<Base>::FUNC()(os, v);
     }
 
     if (v.isParameter()) {

--- a/include/cppad/cg/operation_node_name_streambuf.hpp
+++ b/include/cppad/cg/operation_node_name_streambuf.hpp
@@ -44,16 +44,16 @@ public:
         if (BUF != nullptr) {
             throw CGException("Only one OperationNodeNameStreambuf can exist at a time in each thread");
         }
-        if (CGOStreamFunc<Base>::FUNC != nullptr) {
+        if (CGOStreamFunc<Base>::FUNC() != nullptr) {
             throw CGException("CGOStreamFunc<Base>::FUNC already defined in this thread");
         }
         BUF = this;
-        CGOStreamFunc<Base>::FUNC = registerNode;
+        CGOStreamFunc<Base>::FUNC() = registerNode;
     }
 
     virtual ~OperationNodeNameStreambuf() {
         BUF = nullptr;
-        CGOStreamFunc<Base>::FUNC = nullptr;
+        CGOStreamFunc<Base>::FUNC() = nullptr;
     }
 
     std::streamsize xsputn(const char_type* s,

--- a/test/cppad/cg/CMakeLists.txt
+++ b/test/cppad/cg/CMakeLists.txt
@@ -22,6 +22,7 @@ add_cppadcg_test(inputstream.cpp)
 add_cppadcg_test(temporary.cpp)
 add_cppadcg_test(mult_sparsity_pattern.cpp)
 add_cppadcg_test(multi_object_1.cpp multi_object.cpp)
+add_cppadcg_test(gcc_build_issue.cpp)
 
 ADD_SUBDIRECTORY(extra)
 ADD_SUBDIRECTORY(operations)

--- a/test/cppad/cg/gcc_build_issue.cpp
+++ b/test/cppad/cg/gcc_build_issue.cpp
@@ -1,0 +1,41 @@
+/* --------------------------------------------------------------------------
+ *  CppADCodeGen: C++ Algorithmic Differentiation with Source Code Generation:
+ *    Copyright (C) 2012 Joris Vaillant
+ *
+ *  CppADCodeGen is distributed under multiple licenses:
+ *
+ *   - Eclipse Public License Version 1.0 (EPL1), and
+ *   - GNU General Public License Version 3 (GPL3).
+ *
+ *  EPL1 terms and conditions can be found in the file "epl-v10.txt", while
+ *  terms and conditions for the GPL3 can be found in the file "gpl3.txt".
+ * ----------------------------------------------------------------------------
+ * Author: Joris Vaillant
+ */
+#include "CppADCGTest.hpp"
+
+using namespace CppAD;
+using namespace CppAD::cg;
+
+// Test GCC compilation bug issue: https://github.com/joaoleal/CppADCodeGen/issues/92
+// This test only have to build to pass
+
+struct A {};
+struct B {};
+
+template <typename T>
+struct Wrapper {
+  virtual void foo() { 
+    v.FUNC();
+  }
+  CppAD::cg::CGOStreamFunc<T> v;
+};
+
+int main() {
+};
+
+TEST_F(CppADCGTest, TestGCCBuildIssue) {
+  Wrapper<A> a;
+  Wrapper<B> b;
+  a.v.FUNC();
+}

